### DISCUSSION TITLE
Challenge 1 and 2: Implement `Insert` and `Search` functions for trie

### DISCUSSION THREAD

**Discussion by thomasy314** [ishtms/learn-nodejs-hard-way#53] [General]
I decided to go with an iterative approach as it seemed better for:

1. I felt this was easier to read than a recursive implementation
2. Makes the insert and search functions similar, making it easier to understand both by knowing one
3. Not really as important, but this will reduce load on the call stack. Since words are generally pretty short, I think this is a marginal optimization.

```js
class Trie {
    constructor() {
        this.root = new TrieNode();
    }

    /**
     * Adds a given word to the trie structure 
     * @param {string} word 
     */
    insert(word) {
        let curNode = this.root;
        for (const letter of word) {
            if (letter in curNode.children) {
                curNode = curNode.children[letter];
            } else {
                const newNode = new TrieNode()
                curNode.children[letter] = newNode;
                curNode = newNode;
            }
        }
        curNode.isEndOfWord = true;
    }

    /**
     * Traverses the trie and determines if it contains the given word
     * @param {string} word 
     */
    search(word) {
        let curNode = this.root;
        for (const letter of word) {
            if (letter in curNode.children) {
                curNode = curNode.children[letter];
            } else {
                return false;
            }
        }
        return curNode.isEndOfWord;
    }
}

class TrieNode {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}
```

Happy to hear any feedback or thoughts on my implementation :)

**Comment by ishtms** [comment]
I personally prefer an iterative approach over recursion. I chose to implement the initial solution recursively to highlight the differences and help others appreciate how straightforward it is to create and understand loop-based implementations compared to recursive ones.

I've also included a loop based implementation [in a commit](https://github.com/ishtms/learn-nodejs-hard-way/blob/master/chapters/ch07-ex-implementing-a-trie.md#using-a-for-loop-instead-of-recursion) yesterday.
